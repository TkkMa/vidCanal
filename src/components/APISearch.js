const axios = require('axios');
const moment = require('moment');

const ROOT_URL = {
    YT: 'https://www.googleapis.com/youtube/v3/search',
    D: 'https://api.dailymotion.com/videos',
    V: 'https://api.vimeo.com/videos'
}
const ROOT_URL_VIDEO = {
    YT: 'https://www.googleapis.com/youtube/v3/videos'
    // D: 'https://api.dailymotion.com/video/',
    // V: 'https://api.vimeo.com/videos'
};
const API_KEY= {
    YT:'AIzaSyDWU_q-QSxsSUECDdeWI8WBnVE2q07p804',
    V: 'f0ca0fbcd4918f82208f732dfded4341'
}

const vidAPISearch = (options={}, engine='YT')=>{
    let params;
    if (engine === 'YT'){
        params = {
            maxResults: options.num,
            part: 'snippet',
            key: API_KEY.YT,
            q: options.term,
            type: 'video',
            order: options.sortByVal,
            pageToken: options.pageToken[engine]
        };
        if(options.sortTimeVal){params.publishedAfter=options.sortTimeVal;}
        return axios.get(ROOT_URL[engine], { params: params })
            .then(response=>{           
                return {
                    pageInfo: response.data.pageInfo, 
                    items: response.data.items, 
                    nextPageToken: response.data.nextPageToken || ''
                };
            })
            .catch(function(error) {
                console.error(error);
                return null;
            });
    } else if(engine==='D'){
        params = {
            limit: options.num,
            search: options.term,
            sort: options.sortByVal,
            page: options.pageToken[engine],
            fields: 'id,title,owner.screenname,owner.url,views_total,embed_url,created_time,description,thumbnail_60_url,thumbnail_120_url,thumbnail_180_url'
          };

        if(options.sortTimeVal){params.created_after=moment(options.sortTimeVal).unix();}
        if(!params.sort){console.error('No alphabetical sorting exists for Dailymotion'); return null}
        return axios.get(ROOT_URL[engine], {params: params})
            .then(response=>{
                const arrayOfObj = response.data.list.map(({
                    ['owner.screenname']: ownerScreenname,
                    ['owner.url']: ownerUrl,
                    ...rest
                })=>({
                    ownerScreenname,
                    ownerUrl,
                    ...rest
                }))
                return {
                    pageInfo: {
                        total: response.data.total,
                        has_more: response.data.has_more
                    }, 
                    items: arrayOfObj, 
                    nextPageToken: options.pageToken[engine]+1
                };            
            })
            .catch(function(error) {
                console.error(error);
                return null;
            });
    } else if(engine==='V'){
        params = {
            access_token: API_KEY.V,
            per_page: options.num,
            query: options.term,
            page: options.pageToken[engine],
            sort: options.sortByVal,
            fields:'total,page,per_page,paging,data,uri,pictures,name,release_time,user,stats,description,embed'
        }

        // if(options.sortTimeVal){params.filter=moment(options.sortTimeVal).unix();}
        return axios.get(ROOT_URL[engine], { params: params })
            .then(response=>{
                return {
                    pageInfo: {
                        paging: response.data.paging,
                        total: response.data.total
                    }, 
                    items: response.data.data, 
                    nextPageToken: options.pageToken[engine]+1
                };
            })
            .catch(function(error) {
                console.error(error);
                return null;
            });
    }  
}

//-- Daily motion and Vimeo only requires one API Call in the previous search request to get all fields
//-- Youtube requires two API calls
const indVidAPISearch = (options, engine='YT') =>{
    let params;
    if(engine==='YT'){
        params={
            id: options.id,
            key: API_KEY.YT,
            part: 'snippet, statistics'
        }

        return axios.get(ROOT_URL_VIDEO[engine], {params: params})
            .then(response=>{
            return response.data.items
            })
            .catch(function(error) {
            console.error(error);
        });
    } else if(engine==='D'){
        return [options];
    } else if(engine==='V'){
        return [options];
    }
};

module.exports = {
    vidAPISearch,
    indVidAPISearch
}