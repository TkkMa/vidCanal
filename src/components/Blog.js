import React from 'react';
import { Link } from 'react-router-dom';

const Blog = () => (
  <div className="construction">
    <img src='/images/under-construction-image.png' />
    <Link to="/">Go home</Link>
  </div>
);

export default Blog;