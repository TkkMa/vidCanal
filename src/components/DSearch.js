const axios = require('axios');
const ROOT_URL = 'https://api.dailymotion.com/videos';
// const ROOT_URL_VIDEO = 'https://www.googleapis.com/youtube/v3/videos';
// const API_KEY='AIzaSyDWU_q-QSxsSUECDdeWI8WBnVE2q07p804';

const DSearch = (options={})=>{

  const params = {
    limit: options.num,
    search: options.term,
    sort: options.sortBy,
    page: options.pageToken
  };
  if(options.sortTimeVal){params.createdAfter=options.sortTimeVal;}

  return axios.get(ROOT_URL, { params: params })
    .then((response)=>{
      return {
        has_more: response.data.has_more,
        items: response.data.list, 
        nextPageToken: response.data.page + 1
      };
    })
    .catch(function(error) {
      console.error(error);
    });  
}

const YTVideo = (options) =>{
  const params={
    id: options.id,
    key: API_KEY,
    part: 'snippet, statistics'
  }

  return axios.get(ROOT_URL_VIDEO, {params: params})
    .then((response)=>{
      return response.data.items
    })
    .catch(function(error) {
      console.error(error);
    });
}

module.exports = {
  YTSearch,
  YTVideo
}