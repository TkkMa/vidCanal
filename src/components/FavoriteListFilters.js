import React, {Component} from 'react';
import {connect} from 'react-redux';
import favorites from '../selectors/favorites';
import {startSaveVideo} from '../actions/videos';
import {clearFavCount} from '../actions/filters';
import DeletionModal from './DeletionModal';

class FavoriteListFilters extends Component{

    state ={
        showModal: false,
        message: 'favorite videos'
    };

    componentDidMount(){
        $(".dropdown-trigger").dropdown();
        $('.sidenav').sidenav({
            preventScrolling: false
        });
    };

    handleCloseModal = () =>{
        this.setState({showModal: false});
    };

    handleOpenModal = () =>{
        if (this.props.uniqueLikedVids.length) {this.setState({showModal: true});}
    };
    
    clearFavorites=()=>{
        const dbIds = [];
        const updatedVideos = this.props.visitedVideos.map(video =>{
            if(video.isSaved===true){
                video.isSaved = false
                dbIds.push(video.DB_id);
            }
            return video;
        })
        this.props.startSaveVideo({
            dbIds,
            updatedVideos
        })
        this.props.setFavCount({
            count: 0,
            countIds : []
        });
        this.handleCloseModal();                 
    };

    markAsSeen=()=>{
        this.props.clearFavCount();
    };

    render(){
        return(
            <div>
                <ul id="dropdown1" className="dropdown-content">
                    <li onClick={this.markAsSeen}><a><i className="material-icons">remove_red_eye</i>Seen All</a></li>
                    <li onClick={this.handleOpenModal}><a><i className="material-icons">clear_all</i>Clear</a></li>
                </ul>
                <nav className="FLF-1">
                    <div className="nav-wrapper row">
                        <div className="col s5 m4 l3 xl2 right">
                            <a class="dropdown-trigger" href="#!" data-target="dropdown1"><i class="material-icons left">edit</i>Actions<i class="material-icons right">arrow_drop_down</i></a>
                        </div>
                        <div className="col s1 m1 right">
                            <i className="material-icons">star</i>{this.props.uniqueLikedVids.length}
                        </div>
                        <div className="col s2 m1 l1 xl1 right">
                            <button data-target="slide-out" className="sidenav-trigger btn btn-flat">
                                <i className="material-icons">list</i>
                                {(this.props.searchKeyFilter.length)? <i className="material-icons">check_circle</i>: <i/>}
                            </button>
                        </div>
                    </div>
                </nav>
                <DeletionModal
                    showModal={this.state.showModal}
                    handleCloseModal={this.handleCloseModal}
                    clearList={this.clearFavorites}
                    message={this.state.message}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    searchKeyFilter: state.filters.searchKeyFilter,
    uniqueLikedVids: favorites(state.videos.visitedVideos),
    visitedVideos: state.videos.visitedVideos
});

const mapDispatchToProps = (dispatch)=>({
    startSaveVideo: (video) => dispatch(startSaveVideo(video)),
    clearFavCount: () => dispatch(clearFavCount())
});

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteListFilters);