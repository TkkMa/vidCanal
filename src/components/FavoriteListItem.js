import React, {Component} from "react";
import {connect} from 'react-redux';
import moment from "moment";
import {selectVideo, startSaveVideo} from "../actions/videos";
import {removeFavCount} from '../actions/filters';
import {playerLabelsObj} from '../fixtures/playerLabels';

export class FavoriteListItem extends Component{

    updateCountData = () =>{
        const index = this.props.ids.findIndex(idElement=> idElement.videoId===this.props.video.id);
        if(index>-1){
            this.props.removeFavCount({
                count: this.props.count - 1,
                videoIds: this.props.ids.filter(idElement => idElement.videoId !== this.props.video.id),
                foundDbId: this.props.ids[index].DB_id
            });            
        };
    };

    handleOpenModal = () =>{
        this.updateCountData();
        this.props.handleOpenModal(this.props.video)
    };

    clearItem =(e)=>{
        e.stopPropagation();
        this.updateCountData();
        const updatedVideos = this.props.visitedVideos.map(video=>{
            if (video.id === this.props.video.id){
                return {...video, isSaved:false}
            }
            return {...video};
        });
        this.props.startSaveVideo({
            updatedVideos,
            dbIds: [this.props.video.DB_id]
        });
    };

    render(){
        const {video} = this.props;

        return(
            <li className="collection-item avatar FLI-1" onClick={this.handleOpenModal}>
                <div className="row">
                    <div className="col s12 m12 l3 xl2 div-record-1">
                        <div>
                            <span className="span-search-text">Search term: </span>
                            <span> {video.searchKey}</span>
                        </div>
                        <div>
                            <span className="span-search-text">Last Visited Time:</span>
                            <span> {moment(video.viewedAt).local().format('ddd, MMM D YYYY')}</span>
                            <span> {moment(video.viewedAt).local().format('h:mm:ss a')}</span>    
                        </div>
                    </div>
                    <div className="col s12 m3 l3 xl3 div-record-2">
                        <img src={video.imageUrl_medium} className="responsive-img"/>
                    </div>
                    <div className="col s10 m7 l5 xl6 div-record-3">
                        <span className="title">
                            {video.title}
                            {(this.props.ids.findIndex(idElement=>idElement.videoId===video.id)>-1) ? <span className="new badge"></span>: <span/>}
                        </span>
                        <p>{moment(video.publishedAt).format('DD MMM YYYY')} - {video.viewCount} views<br />
                            Uploaded by:<a href={video.channelUrl}> {video.channelTitle}</a>
                        </p>
                        <p>{video.description.substring(0,200)}...</p>                    
                    </div>
                    <div className="col s2 m2 l1 xl1 div-record-4">
                        <img className="responsive-img" src={playerLabelsObj[video.engine].img} /> 
                        <i className="material-icons right" onClick={this.clearItem}>clear</i>
                    </div>
                </div>
            </li>
        )
    }
}
const mapStateToProps = (state)=>({
    count: state.filters.unViewedFavCount,
    ids: state.filters.unViewedFavIds,
    visitedVideos: state.videos.visitedVideos
})

const mapDispatchToProps = (dispatch)=>({
    selectVideo: (video)=> dispatch(selectVideo(video)),
    removeFavCount: (favCount) => dispatch(removeFavCount(favCount)),
    startSaveVideo: (video) => dispatch(startSaveVideo(video))
});

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteListItem);