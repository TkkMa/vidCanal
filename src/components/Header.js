import React, {Component} from 'react';
import {NavLink, withRouter} from 'react-router-dom';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {setSearchKey} from '../actions/videos';
import SearchBar from './SearchBar';

export class Header extends Component {

    state={
        term: ''
    }

    static propTypes={
        history: PropTypes.object.isRequired
    }

    onSearchChange = (e) =>{
        this.setState({term:e.target.value});
    }

    onSearchSubmit = (e)=>{
        e.preventDefault();
        console.log('query in header: ', this.state.term);
        // $("input[type='text']").val('');        
        this.props.setSearchKey({
            text: this.state.term,
            reRender: true,
            didMount: true
        });
        this.setState({term:''});
        //-- In the case when the search is conducted outside of path '/'.  ComponentWillReceiveProps will not
        //-- be initialised in that component with route '/'        
        if(this.props.location.pathname !== '/'){
            this.props.history.push('/');
        };
    }

    render(){
        const {term} = this.state;
        return(
            <div>
                <SearchBar
                    value={term}
                    onChange={this.onSearchChange}
                    onSubmit={this.onSearchSubmit}
                />
                <div className="row nav-bar">
                    <div className="col s12">
                        <ul ref="menuTab" className="tabsCustom light">
                            <li className="col s2 tabCustom">
                                <NavLink to="/" activeClassName="active" exact={true}>
                                    <i className="material-icons">featured_video</i>Videos
                                </NavLink>
                            </li>
                            <li className="col s2 tabCustom">
                                <NavLink to="/history" activeClassName="active">
                                    <i className="material-icons">history</i>History
                                </NavLink>
                            </li>
                            <li className="col s2 tabCustom">
                                <NavLink to="/saved" activeClassName="active">
                                    <i className="material-icons">star</i><span>Favorites</span>
                                    {(this.props.count) ? 
                                        (<span className="new badge">{this.props.count}</span>) :
                                        ('')
                                    }
                                </NavLink>
                            </li>
                            <li className="col s2 tabCustom">
                                <NavLink to="/blog" activeClassName="active">
                                <i className="material-icons">note_add</i>Blog
                                </NavLink>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state)=>{
    return{
        count: state.filters.count
    }   
};

const mapDispatchToProps = (dispatch)=>({
    setSearchKey: (text)=> dispatch(setSearchKey(text))
});

// export default Header;
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Header));
// export default connect(mapStateToProps, mapDispatchToProps)(Header);