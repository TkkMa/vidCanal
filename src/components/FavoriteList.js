import React, {Component} from 'react';
import {connect} from 'react-redux';
import FavoriteListItem from './FavoriteListItem';
import FavoriteModal from './FavoriteModal';
import favorites from '../selectors/favorites';
// import {setFavCount} from '../actions/filters';

export class FavoriteList extends Component {

    state ={
        showModal: false,
        clickedVideo: {}
    }

    handleOpenModal= (clickedVideo)=>{
        this.setState({
            showModal: true,
            clickedVideo
        });
    }

    handleCloseModal=()=>{
        this.setState({showModal: false});
    }

    render(){
        const filteredVids = this.props.uniqueLikedVideos;
        return(
                <div className="row">
                    {
                        (filteredVids.length === 0) ? (
                            <p>No favorite videos</p>
                        ):(
                            <div className="col s12">
                                <ul className="collection">
                                    {
                                        filteredVids.map((video, index) =>(
                                            <FavoriteListItem 
                                                {...this.props} 
                                                handleOpenModal={this.handleOpenModal}
                                                key={video.id+index} 
                                                video={video}
                                            />
                                        ))
                                    }
                                </ul>
                            </div>
                        )
                    }
                    <FavoriteModal
                        showModal={this.state.showModal}
                        handleCloseModal={this.handleCloseModal}
                        video={this.state.clickedVideo}
                    />
                </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
      visitedVideos: state.videos.visitedVideos,
      uniqueLikedVideos: favorites(state.videos.visitedVideos)
    };
};

export default connect(mapStateToProps)(FavoriteList);