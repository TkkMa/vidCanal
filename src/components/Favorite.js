import React from 'react';
import FavoriteListFilters from './FavoriteListFilters';
import FavoriteList from './FavoriteList';

const Favorite = ()=>{
    return(
        <div className="container">
            <FavoriteListFilters />
            <FavoriteList />
        </div>
    )
}

export default Favorite