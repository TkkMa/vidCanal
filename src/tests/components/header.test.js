import React from 'react';
import {shallow} from 'enzyme';
import {Header} from '../../components/Header';
import { createMemoryHistory } from 'history';

let props, wrapper, e, value, history;
beforeEach(()=>{
    history= createMemoryHistory('/');
    props={history};
    wrapper = shallow(<Header {...props}/>);
    value = 'redux';
    e = {target:{value}, preventDefault: jest.fn()};
})
test('should render Header correctly', ()=>{
    expect(wrapper).toMatchSnapshot();
    // expect(wrapper.find('h1').text()).toBe(1);
    // const renderer = new ReactShallowRenderer();
    // renderer.render(<Header />);
    // expect(renderer.getRenderOutput()).toMatchSnapshot();
})

test('should render onChange of searchBar', ()=>{
    wrapper.find('SearchBar').prop('onChange')(e);
    expect(wrapper.state('term')).toBe(value);
});

test('should submit search results', ()=>{
    const setSearchKeySpy = jest.fn();
    history = {push: jest.fn()};
    const location= {pathname: '/saved'};
    wrapper = shallow(<Header 
                        {...props}
                        setSearchKey={setSearchKeySpy}
                        history={history}
                        location={location}
                    />);
    wrapper.find('SearchBar').prop('onSubmit')(e);
    expect(setSearchKeySpy).toHaveBeenLastCalledWith({
        text: '',
        reRender: true,
        didMount: true
    })
    expect(wrapper.state('term')).toBe('');
    expect(history.push).toHaveBeenLastCalledWith('/');
})