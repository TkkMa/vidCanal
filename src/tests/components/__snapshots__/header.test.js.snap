// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="container"
    >
      <Connect(HeaderBar) />
    </div>
  </div>
  <div
    className="container"
  >
    <SearchBar
      onChange={[Function]}
      onSubmit={[Function]}
      value=""
    />
    <div
      className="row nav-bar"
    >
      <div
        className="col s12"
      >
        <ul
          className="tabsCustom light"
        >
          <li
            className="col s2 tabCustom"
          >
            <NavLink
              activeClass="active"
              activeClassName="active"
              aria-current="page"
              exact={true}
              to="/"
            >
              <i
                className="material-icons"
              >
                featured_video
              </i>
              <span>
                Videos
              </span>
            </NavLink>
          </li>
          <li
            className="col s2 tabCustom"
          >
            <NavLink
              activeClass="active"
              activeClassName="active"
              aria-current="page"
              to="/history"
            >
              <i
                className="material-icons"
              >
                history
              </i>
              <span>
                History
              </span>
            </NavLink>
          </li>
          <li
            className="col s2 tabCustom"
          >
            <NavLink
              activeClass="active"
              activeClassName="active"
              aria-current="page"
              to="/saved"
            >
              <i
                className="material-icons"
              >
                star
              </i>
              <span>
                Faves
              </span>
            </NavLink>
          </li>
          <li
            className="col s2 tabCustom"
          >
            <NavLink
              activeClass="active"
              activeClassName="active"
              aria-current="page"
              to="/blog"
            >
              <i
                className="material-icons"
              >
                note_add
              </i>
              <span>
                Blog
              </span>
            </NavLink>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
